src/transports/ipc.js

```diff
@@ -77,23 +78,46 @@ function decode(socket, callback) {
     return;
   }
 
-  let { op } = working;
-  let raw;
-  if (working.full === '') {
-    op = working.op = packet.readInt32LE(0);
-    const len = packet.readInt32LE(4);
-    raw = packet.slice(8, len + 8);
-  } else {
-    raw = packet.toString();
-  }
+  accumulatedData.payload = Buffer.concat([accumulatedData.payload, packet]);
 
-  try {
-    const data = JSON.parse(working.full + raw);
-    callback({ op, data }); // eslint-disable-line callback-return
-    working.full = '';
-    working.op = undefined;
-  } catch (err) {
-    working.full += raw;
+  while (accumulatedData.payload.length > 0) {
+    if (accumulatedData.expectedLength === 0) {
+      // We are at the start of a new payload
+      accumulatedData.op = accumulatedData.payload.readInt32LE(0);
+      accumulatedData.expectedLength = accumulatedData.payload.readInt32LE(4);
+      accumulatedData.payload = accumulatedData.payload.subarray(8); // Remove opcode and length
+    }
+
+    if (accumulatedData.payload.length < accumulatedData.expectedLength) {
+      // Full payload hasn't been received yet, wait for more data
+      break;
+    }
+
+    // Accumulated data has the full payload and possibly the beginning of the next payload
+    const currentPayload = accumulatedData.payload.subarray(0, accumulatedData.expectedLength);
+    const nextPayload = accumulatedData.payload.subarray(accumulatedData.expectedLength);
+
+    accumulatedData.payload = nextPayload; // Keep remainder for next payload
+
+    try {
+      callback({
+        op: accumulatedData.op,
+        data: JSON.parse(currentPayload.toString('utf8')),
+      });
+
+      // Reset for next payload
+      accumulatedData.op = undefined;
+      accumulatedData.expectedLength = 0;
+    } catch (err) {
+      // Full payload has been received, but is not valid JSON
+      console.error('Error parsing payload:', err);
```

Comment:
I don't think we want to call `console` methods internally.

I have no idea what exactly we should do here instead however.
Maybe add an `error` key to the object in the callback and match on that to emit some event on the `RPCClient`?
(I have no idea why that function even takes a callback, it looks rather sync to me.)

Reply:
Sounds reasonable. `error` event could be emitted on the `RPCClient` with an invalid response error.

Reply:
I've adjusted the error handling:
https://github.com/discordjs/RPC/blob/4b025cd9c6e3db6688cbe9acdd6d3dc246bf82b7/src/transports/ipc.js#L113

I'm creating a new error to be more descriptive, but the original error is preserved as the [`cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause).